src/compiler/factory/nodeFactory.ts

```diff
@@ -2644,13 +2654,15 @@ export function createNodeFactory(flags: NodeFactoryFlags, baseFactory: BaseNode
     function createImportTypeNode(
         argument: TypeNode,
         assertions?: ImportTypeAssertionContainer,
+        attributes?: ImportTypeAttributes,

```

Comment:
@rbuckton thoughts on the node factory API changes? I think we will want to avoid shifting the parameters like this, and may need an overload here on these that fall in the middle of the signature.

Alternatively, maybe the idea on all of them was to create a deprecated overload with `assertions` and change the non-deprecated overload to take only `attributes`, and ensure we never parse a node with both assertions and attributes?

Reply:
@andrewbranch do you mean to create the overloads directly in `types`/`nodeFactory` (something like in the example below) or include them in `deprecatedCompat`? 

```ts
/** @deprecated */ createImportTypeNode(argument: TypeNode, assertions?: ImportTypeAssertionContainer, qualifier?: EntityName, typeArguments?: readonly TypeNode[], isTypeOf?: boolean): ImportTypeNode;
/** @deprecated */ updateImportTypeNode(node: ImportTypeNode, argument: TypeNode, assertions: ImportTypeAssertionContainer | undefined, qualifier: EntityName | undefined, typeArguments: readonly TypeNode[] | undefined, isTypeOf?: boolean): ImportTypeNode;
createImportTypeNode(argument: TypeNode, attributes?: ImportTypeAttributes, qualifier?: EntityName, typeArguments?: readonly TypeNode[], isTypeOf?: boolean): ImportTypeNode;
updateImportTypeNode(node: ImportTypeNode, argument: TypeNode, attributes: ImportTypeAttributes | undefined, qualifier: EntityName | undefined, typeArguments: readonly TypeNode[] | undefined, isTypeOf?: boolean): ImportTypeNode;
```

```ts
function createImportTypeNode(
    argument: TypeNode,
    assertionsOrAttributes?: ImportTypeAssertionContainer | ImportTypeAttributes,
    qualifier?: EntityName,
    typeArguments?: readonly TypeNode[],
    isTypeOf = false,
): ImportTypeNode {
    const node = createBaseNode<ImportTypeNode>(SyntaxKind.ImportType);
    node.argument = argument;
    if (assertionsOrAttributes) {
        if (assertionsOrAttributes.kind === SyntaxKind.ImportTypeAttributes) {
            node.attributes = assertionsOrAttributes;
        }
        else {
            node.assertions = assertionsOrAttributes;
        }
    }
    
    node.qualifier = qualifier;
    node.typeArguments = typeArguments && parenthesizerRules().parenthesizeTypeArguments(typeArguments);
    node.isTypeOf = isTypeOf;
    node.transformFlags = TransformFlags.ContainsTypeScript;
    return node;
}
```

Reply:
The example above is what I was talking about, but honestly I’m not familiar with `deprecatedCompat` at all. It would be great to get @rbuckton’s advice.

Reply:
`deprecatedCompat` is mostly about patching our public API to allow the use of deprecated functionality, including warnings, without polluting tsc or allowing *us* to accidently depend on deprecated behavior. Except for a few cases where our API design makes it difficult to handle via patching a specific API, most overloads to handle deprecations should be moved to `deprecatedCompat`.

Reply:
> @andrewbranch do you mean to create the overloads directly in `types`/`nodeFactory` (something like in the example below) or include them in `deprecatedCompat`?
> ...

This is definitely the approach we should take (accepting the union on the 2nd argument), since you can't have *both* an `assert {}` and a `with {}` on an import.

I would do this in `compiler`. There's no point in moving it to `deprecatedCompat` if *we* still need to support the syntax (i.e., we still parse it), otherwise we'd issue deprecation warnings in the console against ourselves when you use that syntax in a source file, as opposed to issuing a normal diagnostic message.

Reply:
Another approach we could *maybe* have taken this would be to just add a `token: SyntaxKind.AssertsKeyword | SyntaxKind.WithKeyword` to the existing `AssertClause` and just rename it and add type aliases pointing to it marked as deprecated.

Then we could have done `node.attributes = node.assertions = attributes` in the factory, but existing tools wouldn't really have a different AST node to contend with.

Reply:
The `token` approach is the same thing we do for `HeritageClause` to support both `extends` and `implements`.

Reply:
@rbuckton Thanks for the detailed feedback. I considered renaming the current interfaces and adding aliases to `Import Assertions*`. However, I've decided to introduce new interfaces for `ImportAttributes` due to several differences from `ImportAssertions`. For instance, `ImportAttributeName` can be a NumericLiteral, which isn't allowed in `Import Assertions`, etc. In the following case, should both 'assertions' and 'attributes' reference the same types, or am I mistaken?

```ts
node.attributes = node.assertions = attributes;
```

Reply:
@nicolo-ribaudo Thanks for the update. I didn't know about the latest changes.

Reply:
@a-tarasyuk let’s go ahead and proceed without support numeric literal keys, and we can loosen it later if necessary. I like Ron’s [suggestion](https://github.com/microsoft/TypeScript/pull/54242#discussion_r1329330025) of reusing the existing `AssertClause` node.